### ISSUE TITLE
cmake: curl subdirectory failed build due to getpid within smb.c

### ISSUE THREAD

**Issue by trackpadpro** [curl/curl#8220]
### I did this

Added cURL as a subdirectory within a CMake build. I am using MbedTLS, but that seems unrelated to this error. When I have more time I will likely try other CMake-compatible libraries. I only started to get this error once I started calling the cURL library within my main.cpp. 

### I expected the following

Everything to build. If I did not alter anything, I got the following:
```
[  0%] Building C object deps/curl/lib/CMakeFiles/libcurl.dir/smb.c.o
/c/Users/Valentin/Downloads/stateSpaceSolver/deps/curl/lib/smb.c: In function 'smb_format_message':
/c/Users/Valentin/Downloads/stateSpaceSolver/deps/curl/lib/smb.c:36:16: warning: implicit declaration of function '_getpid'; did you mean 'getpid'? [-Wimplicit-function-declaration]
   36 | #define getpid _getpid
      |                ^~~~~~~
/c/Users/Valentin/Downloads/stateSpaceSolver/deps/curl/lib/smb.c:371:9: note: in expansion of macro 'getpid'
  371 |   pid = getpid();
      |         ^~~~~~
/c/Users/Valentin/Downloads/stateSpaceSolver/deps/curl/lib/smb.c:36:16: warning: nested extern declaration of '_getpid' [-Wnested-externs]
   36 | #define getpid _getpid
      |                ^~~~~~~
/c/Users/Valentin/Downloads/stateSpaceSolver/deps/curl/lib/smb.c:371:9: note: in expansion of macro 'getpid'
  371 |   pid = getpid();
      |         ^~~~~~
[  0%] Linking C shared library libcurl.dll
/usr/lib/gcc/x86_64-pc-msys/11.2.0/../../../../x86_64-pc-msys/bin/ld: CMakeFiles/libcurl.dir/smb.c.o:smb.c:(.text+0x5aa): undefined reference to `_getpid'
collect2: error: ld returned 1 exit status
```

If I simply commented out the definition of getpid as _getpid, then I got no errors, but my libcurl.dll was not found when running the executable. I am fairly certain that my library inclusion is correct within my project's CMakeLists.txt and main.cpp, so I am assuming that there was a build problem caused by this line change. 

### curl/libcurl version

7.81.0-DEV

### operating system

Windows 10 (CMake through MSYS2 [msys64])

**Comment by bagder** [comment]
I'm not sure I understand how this is a curl bug. The cmake build works fine when used as intended as we've designed it as far as I know - and we also verify that fairly well in CI builds. How is your build different than the normal ones?

**Comment by jay** [comment]
[_getpid](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/getpid?view=msvc-170) is the name of the function in the Windows CRT and is declared in its process.h. I think the OP is targeting msys2 which is not native Windows, it is a fork of cygwin. mingw32/mingw64 targets can make native Windows programs. See [here](https://www.msys2.org/wiki/How-does-MSYS2-differ-from-Cygwin/). I guess it is not correct to define it for everything other than MSDOS, it looks like we should limit it to WIN32, for example

~~~diff
diff --git a/lib/smb.c b/lib/smb.c
index fd49cf6..af458e3 100644
--- a/lib/smb.c
+++ b/lib/smb.c
@@ -32,7 +32,7 @@
 #include <process.h>
 #ifdef CURL_WINDOWS_APP
 #define getpid GetCurrentProcessId
-#elif !defined(MSDOS)
+#elif defined(WIN32)
 #define getpid _getpid
 #endif
 #endif
 ~~~

/cc @MarcelRaad

**Comment by trackpadpro** [comment]
I am not sure how exactly I am differing from intended, and I will try out some other configurations that I believe to be more vanilla, but the reason I lean to it being a curl bug instead of an msys bug is the ```undefined reference to `_getpid'``` within curl's smb.c. 

For my build, I am using curl as a subdirectory, which is allegedly supported by what I read online, and it uses mbedtls. It initially had trouble finding the mbedtls package, but I fed it the exact paths to the relevant MbedTLS libraries. Library linking to my project is not the initial issue (the reason I noted that libcurl.dll was not found is due to my quick fix attempt for the actual bug). As far as I can see, the warnings and subsequent error are caused by the interaction between cmake and gcc in msys2 building libcurl.

And yes, I did think the bug was related to the MSDOS condition, but I also do not know why my library is still corrupted after that unless it is due to another warning which I can send at a later time.

**Comment by trackpadpro** [comment]
```
/c/Users/Valentin/Downloads/stateSpaceSolver/deps/curl/lib/if2ip.c: In function 'Curl_if2ip':
/c/Users/Valentin/Downloads/stateSpaceSolver/deps/curl/lib/if2ip.c:213:19: warning: overflow in conversion from 'long int' to 'int' changes value from '2152756070' to '-2142211226' [-Woverflow]
  213 |   if(ioctl(dummy, SIOCGIFADDR, &req) < 0) {
      |
```
Here is the other warning. Did not think it was related, but it could be. Was not sure if the conversion was meant to cause a sign inversion or if it also had something to do with a curl incompatibility with my configuration.

**Comment by jay** [comment]
long is 8 bytes LP64 for 64-bit msys2 (and cygwin) targets. SIOCGIFADDR is type long constant. ioctl prototype is `int __cdecl ioctl (int __fd, int __cmd, ...);` and `printf("%ld\n", SIOCGIFADDR);` shows 2152756070 (which is > INT_MAX). However, I don't get warnings when I compile if2ip.c for msys2 or cygwin with -Woverflow. (Note: I did have to undefine HAVE_GETIFADDRS to get the mentioned section of code included).

I also tried the self contained example below in both cygwin and msys2 and get no warnings compiling with `gcc -o a a.c -Wall -Wextra -Woverflow`
~~~c
int func(int x)
{
  printf("%d\n", x);
  return 0;
}

int main(int argc, char *argv[]){
  (void)argc;
  long x = atol(argv[1]);
  func(x);
  func(2152756070L);
  return 0;
}
~~~

Seems like I should get warning, but I don't.

**Comment by bagder** [comment]
Also, we build with msys2 in the CI and those builds do just fine. [recent example log](https://dev.azure.com/daniel0244/curl/_build/results?buildId=8204&view=logs&jobId=ea1d4c80-679f-50fc-e5c4-6cbed7e43517&j=ea1d4c80-679f-50fc-e5c4-6cbed7e43517&t=6dc28ef7-2393-5edf-bed4-5e258baa3367)

**Comment by trackpadpro** [comment]
Compiler info: 
```
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Updating submodules
-- curl version=[7.81.0-DEV]
```
Not entirely sure how to check if CMake is running -Wconversion but I certainly did not manually toggle that on. Either way, as I said before, it this probably unrelated to the getpid issue.

**Comment by trackpadpro** [comment]
https://github.com/trackpadpro/curl/actions/runs/1659756839

Okay, I figured it out. My problem was I had my project bootstrap using "make" instead of using ninja or mingw32-make. I went back and re-read the install document and I am probably completely misunderstanding how to use what. Make should only be used with sh configure? Not entirely sure what that means, but I will close this issue because I can just use mingw32 makefiles instead of msys makefiles.

**Comment by bagder** [comment]
> Make should only be used with sh configure?

Not at all. You can have cmake generate makefiles and then use make.

**Comment by trackpadpro** [comment]
To clarify, though, CMake would still have to be choosing mingw32 makefiles to be necessarily supported, right?

Thanks

**Comment by bagder** [comment]
Sorry, I don't know much cmake and know even less about using it on Windows. I was only talking from my experience of using cmake on Linux. I have no idea what using it with msys/mingw requires or conditions.

**Comment by orgads** [comment]
It happens with cmake, cannot reproduce with configure/make. Proposed a fix: #9255
